src/h3/mod.rs

```diff
@@ -1125,42 +1153,73 @@ impl Connection {
             return Ok((finished, Event::Finished));
         }
 
-        // Process DATAGRAMs
+        if self.dgram_event_count == self.dgram_poll_threshold &&
+            self.stream_event_count == self.stream_poll_threshold
+        {
+            self.dgram_event_count = 0;
+            self.stream_event_count = 0;
+        }
         let mut d = [0; 8];
 
-        match conn.dgram_recv_peek(&mut d, 8) {
-            Ok(_) => {
-                let mut b = octets::Octets::with_slice(&d);
-                let flow_id = b.get_varint()?;
-                return Ok((flow_id, Event::Datagram));
-            },
+        if self.dgram_event_count < self.dgram_poll_threshold &&
+            conn.dgram_recv_front_len().is_some()
+        {
+            match conn.dgram_recv_peek(&mut d, 8) {
```

Comment:
Might be a good idea to refactor this into a separate method, so it can be reused later rather than duplicating the code, though I'm not too sure it would actually simplify things, as you'd still need the `match` here anyway.

Reply:
Good idea, see `process_dgrams() which still returns a Result but handles the boring mapping between error types, which makes the code in `poll()` a little tidier.